### PR TITLE
remove apply_annotations from solver module

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**

### CODE REVIEW

File: tests/test_solver.py [hunk 1]
```diff
@@ -14,15 +14,38 @@ def tmpcache():
...
+        else v) # convert arrays to lists so we can use assertEqual
+
+
 def _test_recursion_cache(testcase, solver_iter):
-    edit = lambda v: v.tolist() if numeric.isarray(v) else tuple(map(edit, v)) if isinstance(v, tuple) else v  # convert arrays to lists
-    read = lambda n: tuple(edit(item) for i, item in zip(range(n), solver_iter()))
+    read = lambda n: tuple(_edit(item) for i, item in zip(range(n), solver_iter))
     reference = read(5)
-    for lengths in [1, 2, 3], [1, 3, 2], [0, 3, 5]:
+    for lengths in [1, 2, 3], [1, 3, 2], [3, 5]:
```
Comment (joostvanzwieten) [inline tests/test_solver.py:27]: Why is `0` removed from the last set of lengths?

File: nutils/solver.py [hunk 2]
```diff
@@ -394,8 +388,12 @@ class newton(cache.Recursion, length=1):
         (boolean) or NaN (float). In the remaining positions the values of
         ``lhs0`` are returned unchanged (boolean) or overruled by the values in
         `constrain` (float).
-    linesearch : :class:`nutils.solver.LineSearch`
-        Callable that defines relaxation logic.
+    linesearch : Callable[(float, float, float, float), float]
```
Comment (joostvanzwieten) [inline nutils/solver.py:272]: I think the return value is incorrect. Shouldn't this be `Tuple[float, bool]`? Also, in the [docs for `Callable`](https://docs.python.org/3/library/typing.html#typing.Callable) a list is used as first argument, not a tuple. The docs don't really state this as a requirement, though.

File: nutils/solver.py [hunk 3]
```diff
@@ -794,8 +792,12 @@ def optimize(target, functional: evaluable.asarray, *, tol: types.strictfloat =
         Coefficient vector, starting point of the iterative procedure.
     relax0 : :class:`float`
         Initial relaxation value.
-    linesearch : :class:`nutils.solver.LineSearch`
-        Callable that defines relaxation logic.
+    linesearch : Callable[(float, float, float, float), float]
```
Comment (joostvanzwieten) [inline nutils/solver.py:716]: Same as above.

File: CHANGELOG [hunk 4]
```diff
@@ -5,6 +5,17 @@ The following overview lists user facing changes as well as newly added
 features in inverse chronological order.
 
 
+CHANGED: solve, solve_withinfo arguments
+
+Solver methods newton, minimize and pseudotime have their function signature
+slightly changed: 1. The tol argument (used to define the stop criterion) has
+been bade mandatory. As the default value used to be 0 - an unreacheable value
```
Comment (joostvanzwieten) [inline CHANGELOG:12]: `bade` -> `made`

File: tests/test_solver.py (hunk 1 above)
Comment (gertjanvanzwieten) [reply tests/test_solver.py]: restored, along with lambda functions

File: nutils/solver.py (hunk 2 above)
Comment (gertjanvanzwieten) [reply nutils/solver.py]: fixed both issues

File: nutils/solver.py (hunk 3 above)
Comment (gertjanvanzwieten) [reply nutils/solver.py]: fixed both issues

File: CHANGELOG (hunk 4 above)
Comment (gertjanvanzwieten) [reply CHANGELOG]: fixed

File: nutils/solver.py (hunk 2 above)
Comment (joostvanzwieten) [reply nutils/solver.py]: The return type is now `bool`, not `Tuple[float, bool]`.

Comment (gertjanvanzwieten) [reply nutils/solver.py]: fix for real this time

**Review by joostvanzwieten — APPROVED** [review]

**Pull request merged**